socorro/external/es/crashstorage.py

```diff
@@ -87,185 +68,219 @@ def is_valid_key(key):
 
-        value_bytes = value_bytes[:max_size]
-        new_value = ""
+    :param value: the value to fix
 
-        # Remove bytes until we either run out of bytes or we can decode to a unicode
-        # string
-        while value_bytes:
-            try:
-                new_value = value_bytes.decode("utf-8")
-                break
-            except UnicodeDecodeError:
-                value_bytes = value_bytes[:-1]
+    :returns: fixed value
 
-        new_value = new_value or "BAD DATA"
+    """
+    min_int, max_int = INTEGER_BOUNDS
+    if not isinstance(value, int):
+        try:
+            value = int(value)
+        except ValueError:
+            # If the value isn't valid, remove it
+            return None
 
-        data[field_name] = new_value
+    if not (min_int <= value <= max_int):
+        # If the value isn't within the bounds, remove it
+        return None
 
+    return value
 
-POSSIBLE_TRUE_VALUES = [1, "1", "true", True]
 
+LONG_BOUNDS = (-9_223_372_036_854_775_808, 9_223_372_036_854_775_807)
 
-def convert_booleans(fields, data):
-    """Converts pseudo-boolean values to boolean values for boolean fields
 
-    Valid boolean values are True, 'true', False, and 'false'.
+def fix_long(value):
+    """Fix long value so it doesn't exceed Elasticsearch maximums
 
-    Note: This modifies the data dict in-place and only looks at the top level.
+    :param value: the value to fix
 
-    :arg dict fields: the super search fields schema
-    :arg dict data: the data to look through
+    :returns: fixed value
 
     """
-    boolean_fields = get_fields_by_item(fields, "analyzer", "boolean")
+    min_long, max_long = LONG_BOUNDS
+    if not isinstance(value, int):
+        try:
+            value = int(value)
+        except ValueError:
+            # If the value isn't valid, remove it
+            return None
 
-    for field in boolean_fields:
-        field_name = field["in_database_name"]
+    if not (min_long <= value <= max_long):
+        # If the value isn't within the bounds, remove it
+        return None
+    return value
 
-        value = data.get(field_name)
-        data[field_name] = True if value in POSSIBLE_TRUE_VALUES else False
 
+def fix_datetime(value):
+    """Fix datetime value to index correctly
 
-INTEGER_BOUNDS = (-2_147_483_648, 2_147_483_647)
-LONG_BOUNDS = (-9_223_372_036_854_775_808, 9_223_372_036_854_775_807)
+    :param value: the value to fix
 
+    :returns: fixed value
 
-def fix_numbers(fields, data):
-    """Fix number values so they don't exceed Elasticsearch maximums
+    """
+    if isinstance(value, str):
+        return string_to_datetime(value)
+    if not isinstance(value, (datetime.datetime, datetime.time)):
+        return None
+    return value
 
-    "long" can be -9,223,372,036,854,775,808
 
-    :arg dict fields: the super search fields schema
-    :arg dict data: the data to look through
+def build_document(src, crash_document, fields, all_keys):
```

Comment:
This way of building the document is much better than the previous way in that it's much easier to reason about, iterates over fields once, and makes the related code easier to test and modify over time.